iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java

```diff
@@ -218,10 +221,28 @@ public void purgeSoftDeletedTables(String databaseId, String tableId, long purge
         databaseId,
         tableId,
         purgeAfterMs);
-
     houseTableRepository.purgeSoftDeletedTables(databaseId, tableId, purgeAfterMs);
   }
 
+  public void restoreTable(String databaseId, String tableId, long deletedAtMs) {
+    log.info(
+        "Restoring soft deleted table for databaseId: {}, tableId: {}, deletedAtMs: {}",
+        databaseId,
+        tableId,
+        deletedAtMs);
+    try {
+      houseTableRepository.restoreTable(databaseId, tableId, deletedAtMs);
+    } catch (HouseTableNotFoundException e) {
+      throw new NoSuchUserTableException(
+          databaseId,
+          tableId,
+          String.format("Soft deleted user table not found with timestamp %s", deletedAtMs),
+          e);
+    } catch (HouseTableConcurrentUpdateException e) {
+      throw new AlreadyExistsException("Table", databaseId + "." + tableId, e);
```

Comment:
Does this mean table has been already restored?

Reply:
It can be any scenario where the table name is currently exists.
E.g. I dropped table <db>.<T> and recreated it as <db>.<T'>. After a day I realize that I have corrupted data and want to roll back to a soft deleted version. I try to restore a table but I needed to have dropped <db>.<T'> first